### PR TITLE
Add BGPPeer VRF support

### PR DESCRIPTION
**Pull request by fedepaol**
Here we add a VRF optional field to the BGPPeer to indicate the session we want to establish with the given peer must happen from an interface belonging to a linux VRF

### CODE REVIEW

File: internal/bgp/frr/frr_bfd_test.go [hunk 1]
```diff
@@ -121,7 +122,20 @@ func TestBFDWithSession(t *testing.T) {
...
+			PeerAddress:   "10.2.2.254:179",
+			SourceAddress: net.ParseIP("10.1.1.254"),
+			MyASN:         100,
+			RouterID:      net.ParseIP("10.1.1.254"),
+			PeerASN:       200,
+			HoldTime:      time.Second,
+			KeepAliveTime: 2 * time.Second,
+			Password:      "password",
+			CurrentNode:   "hostname",
+			EBGPMultiHop:  false,
```
Comment (liornoy) [inline internal/bgp/frr/frr_bfd_test.go:136]: EBGPMultiHop was true before.

File: internal/bgp/frr/frr_test.go [hunk 2]
```diff
@@ -214,17 +316,42 @@ func TestTwoIPv6Sessions(t *testing.T) {
...
+			PeerAddress:   "[10:4:4::255]:179",
+			SourceAddress: net.ParseIP("10:3:3::254"),
+			MyASN:         300,
+			RouterID:      net.ParseIP("10.1.1.254"),
+			PeerASN:       400,
+			HoldTime:      time.Second,
+			KeepAliveTime: time.Second,
+			Password:      "password",
+			CurrentNode:   "hostname",
+			EBGPMultiHop:  true,
```
Comment (liornoy) [inline internal/bgp/frr/frr_test.go:348]: EBGPMultiHop was false.

File: internal/bgp/frr/frr_test.go [hunk 3]
```diff
@@ -274,7 +452,19 @@ func TestSingleAdvertisement(t *testing.T) {
...
+			PeerAddress:   "10.2.2.254:179",
+			SourceAddress: net.ParseIP("10.1.1.254"),
+			MyASN:         100,
+			RouterID:      net.ParseIP("10.1.1.254"),
+			PeerASN:       200,
+			HoldTime:      time.Second,
+			KeepAliveTime: time.Second,
+			Password:      "password",
+			CurrentNode:   "hostname",
+			EBGPMultiHop:  false,
```
Comment (liornoy) [inline internal/bgp/frr/frr_test.go:466]: EBGPMultiHop was false.

File: internal/bgp/frr/frr_test.go [hunk 4]
```diff
@@ -309,7 +499,20 @@ func TestSingleAdvertisementNoRouterID(t *testing.T) {
...
+			PeerAddress:   "10.2.2.254:179",
+			SourceAddress: net.ParseIP("10.1.1.254"),
+			MyASN:         100,
+			RouterID:      nil,
+			PeerASN:       200,
+			HoldTime:      time.Second,
+			KeepAliveTime: time.Second,
+			Password:      "password",
+			CurrentNode:   "hostname",
+			EBGPMultiHop:  false,
```
Comment (liornoy) [inline internal/bgp/frr/frr_test.go:513]: EBGPMultiHop was true.

File: internal/bgp/frr/frr_test.go [hunk 5]
```diff
@@ -338,7 +541,20 @@ func TestSingleAdvertisementInvalidNoPort(t *testing.T) {
...
+			PeerAddress:   "10.2.2.254",
+			SourceAddress: net.ParseIP("10.1.1.254"),
+			MyASN:         100,
+			RouterID:      net.ParseIP("10.1.1.254"),
+			PeerASN:       200,
+			HoldTime:      time.Second,
+			KeepAliveTime: time.Second,
+			Password:      "password",
+			CurrentNode:   "hostname",
+			EBGPMultiHop:  false,
```
Comment (liornoy) [inline internal/bgp/frr/frr_test.go:555]: EBGPMultiHop was true.

File: internal/bgp/frr/frr_test.go [hunk 6]
```diff
@@ -353,7 +569,20 @@ func TestSingleAdvertisementStop(t *testing.T) {
...
+			PeerAddress:   "10.2.2.254:179",
+			SourceAddress: net.ParseIP("10.1.1.254"),
+			MyASN:         100,
+			RouterID:      net.ParseIP("10.1.1.254"),
+			PeerASN:       200,
+			HoldTime:      time.Second,
+			KeepAliveTime: time.Second,
+			Password:      "password",
+			CurrentNode:   "hostname",
+			EBGPMultiHop:  false,
```
Comment (liornoy) [inline internal/bgp/frr/frr_test.go:583]: EBGPMultiHop was false.

File: internal/bgp/frr/frr_test.go [hunk 7]
```diff
@@ -387,7 +616,19 @@ func TestSingleAdvertisementChange(t *testing.T) {
...
```
Comment (liornoy) [inline internal/bgp/frr/frr_test.go:630]: EBGPMultiHop was true.

File: internal/bgp/frr/frr_test.go [hunk 8]
```diff
@@ -430,7 +671,20 @@ func TestSingleAdvertisementWithPeerSelector(t *testing.T) {
...
```
Comment (liornoy) [inline internal/bgp/frr/frr_test.go:686]: EBGPMultiHop was true.

File: internal/bgp/frr/frr_vrf_test.go [hunk 9]
```diff
@@ -0,0 +1,613 @@
...
+		t.Fatalf("Could not create session: %s", err)
+	}
+	defer session1.Close()
+	session2, err := sessionManager.NewSession(l,
+		bgp.SessionParameters{
+			PeerAddress:   "10.2.2.254:179",
+			SourceAddress: net.ParseIP("10.3.3.254"),
+			MyASN:         300,
+			RouterID:      net.ParseIP("10.3.3.254"),
+			PeerASN:       400,
```
Comment (liornoy) [inline internal/bgp/frr/frr_vrf_test.go:150]: If peerAddress is the same for both sessions, shouldn't PeerASN be the same as well?

File: internal/bgp/frr/frr_bfd_test.go (hunk 1 above)
Comment (fedepaol) [reply internal/bgp/frr/frr_bfd_test.go:136]: Yep, copy paste got out of hand :-)

File: internal/bgp/frr/frr_vrf_test.go (hunk 9 above)
Comment (fedepaol) [reply internal/bgp/frr/frr_vrf_test.go:150]: Not necessarily. The same IP reached from two VRF might not (and normally won't, that's the point of using VRF) refer to the same endpoint. So this configuration is totally legit.

**Review by oribon — APPROVED** [review]
apart from some golden files changing their ebgp-multihop this looks good

**Pull request merged**